better-code/src/chapter-2-contracts.md

```diff
@@ -0,0 +1,1044 @@
+contract. Improvements in development speed come from arriving at the
+right design earlier, since a poor experience for readers of your
+documentation becomes a “code smell.”
+
+#### Use Documentation Comments
+
+Contract documentation should always go in your code as comments,
+because:
+
+1. That puts the two things that should correspond—documentation and
+   implementation—next to one another, so you can see when they don't
+   match up. People sometimes complain that docs go out of date, but
+   that's kinda the point: without the ability to see that
+   inconsistency, there's no way to know that there's a bug.
+
+2. Using comments makes it reasonable to combine the activities of
+   coding and documentation, which—believe it or not—are mutually
+   supportive.
+
+3. IDEs and documentation extractors recognize special three-slash
+   syntax to provide an improved developer experience. Using the
+   recognized syntax opens the door to other forms of processing in
+   the future.
+
+#### On Code Review
+
+Contracts also deserve early attention in code reviews.  The *first
+step* should be to review the APIs: all declarations (e.g. function
+signatures) and their associated contracts.  A change that omits
+documentation for any declaration should be sent back for revision,
+because there's no way to make a judgement about that component's
+correctness.
+
+Then, if the APIs aren't *well*-documented (and designed), there's
+little point in looking at their implementations.  Remember, APIs are
+the connective tissue; they're what every client of a component
+interacts with, and will have effects throughout the codebase.  The
+implementation of the API had better be an expression of the contract,
+but deficiencies in the implementation of a function can only do local
+damage.
+
+### A Tower of Invariants
+
+The tower of abstraction mentioned earlier comes with a tower of
+invariants.  The invariants of `PairArray` are built on—and depend
+on—the invariants of the individual arrays. We can embed `PairArray`
+into some larger data structure with its own invariant, which will
+depend on that of `PairArray`.
+
+In fact, you can think of the entire state of your program as one big
+data structure with its own invariants, and formalize what it means
+for the program to “be in a good state” that way.  For example, you
+might have a database of employees, each of which has its own ID, and
+which also stores the ID of the employee's manager (the CEO gets to be
+their own manager).
+
+It's an invariant of your program that a manager ID can't just be
+random; it has to identify an employee that's in the database—that's
+part of what it means for the program to be in a good state, and all
+through the program you have code to ensure that invariant is upheld.
+It would be a great idea to identify and document that whole-program
+invariant in some central location, so maintainers will know to uphold it, and that they
+can rely on it.
```

Comment:
Continuing [this conversation](https://github.com/stlab/better-code/pull/21/files/6a7e2a18b38b4d51a3e93380bedb16cfe4838be3#r2226436603).

I understand where you're coming from, but this seems like a band-aid for poor design. This could be a useful practice for legacy systems where refactoring is costly, but I think the goal should be to design systems with local invariants.

Reply:
Of course, that's the whole point of the next section.  I didn't say it's the _best_ idea, that's coming in the very next sentence.  I feel like you're picking nits, taking one phrase completely out of context.